stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4664,49 +4664,54 @@ LogicalResult ReplicaIdOp::inferReturnTypes(
 // If Op
 //===----------------------------------------------------------------------===//
 
-static LogicalResult verifyConditionalBranch(Operation* op, Region& region,
-                                             llvm::Twine branchName) {
-  if (region.getNumArguments() != 0)
-    return op->emitOpError()
-           << branchName << " must have 0 arguments, but found "
-           << region.getNumArguments();
-
-  TypeRange branchReturnTypes =
-      region.front().getTerminator()->getOperandTypes();
-  if (branchReturnTypes != op->getResultTypes())
-    return op->emitOpError()
-           << branchName << " returned types (" << branchReturnTypes
-           << ") do not match op result types (" << op->getResultTypes() << ")";
-
+static LogicalResult inferConditionalReturnTypeComponents(
+    RegionRange branches, Optional<Location> location,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  if (branches.empty())
+    return emitOptionalError(location, "expect at least one branch");
+
+  ValueTypeRange<OperandRange> branch0ResultTypes =
```

Comment:
I am not sure I understand the logic here. It seems that shape inference will always return the type of branch0. If branch0 returns f32[5] and branch1 returns f32[?] then the correct type to infer is the type of branch1.

Is it true that the correctness criterion for shape inference is that if it succeeds with f32[5] then any execution of the code is guaranteed to have that shape?

Reply:
Hi George, thanks for your comment! Yes a more complicated case would be branch0 returns tensor<5x?xf32> while branch1 returns tensor<?x5xf32>: the ideal expected result should be tensor<5x5xf32>.
This PR provides a basic version to include dynamism in type inference. I added a while ago an util function in [inferMostSpecificType](https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/Base.h#L218) and here is the new PR to fix this: https://github.com/openxla/stablehlo/pull/229

Yes, inferring dim=5 is always true no matter which branch is taken in executions.

Reply:
I would have said that in your example the inferred shape should be <?x?xf32>. I commented in #229.